presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyPredicates.java

```diff
@@ -0,0 +1,143 @@
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(), plan);
+    }
+
+    public static Expression simplifyPredicates(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new SimplifyPredicatesExpressionRewriter(), expression);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = context.rewrite(node.getSource());
+            Expression simplified = simplifyPredicates(node.getPredicate());
+            return new FilterNode(node.getId(), source, simplified);
+        }
+    }
+
+    private static class SimplifyPredicatesExpressionRewriter
+            extends ExpressionRewriter<Void>
+    {
+        @Override
+        public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            treeRewriter.rewrite(node.getLeft(), context);
+            treeRewriter.rewrite(node.getRight(), context);
+
+            List<Expression> predicates = extractPredicates(node.getType(), node);
+
+            List<List<Expression>> subPredicates = getSubPredicates(predicates);
+
+            List<Set<Expression>> deterministicSubPredicates = subPredicates.stream()
+                    .map(p -> p.stream()
+                            .filter(DeterminismEvaluator::isDeterministic)
+                            .collect(toSet()))
+                    .collect(toList());
+
+            Set<Expression> commonPredicates = new HashSet<>(deterministicSubPredicates.stream()
+                    .reduce(Sets::intersection)
+                    .orElse(emptySet()));
+
+            List<List<Expression>> uncorrelatedSubPredicates = subPredicates.stream()
+                    .map(p -> removeAll(p, commonPredicates))
+                    .filter(p -> !p.isEmpty())
+                    .collect(toList());
+
+            if (commonPredicates.isEmpty() || uncorrelatedSubPredicates.isEmpty()) {
+                return combinePredicates(node.getType(), predicates);
+            }
+
+            LogicalBinaryExpression.Type flippedNodeType = flipLogicalBinaryExpressionType(node.getType());
+
+            List<Expression> uncorrelatedPredicates = uncorrelatedSubPredicates.stream()
+                    .map(p -> combinePredicates(flippedNodeType, p))
+                    .collect(Collectors.toList());
+
+            Expression result = combinePredicates(flippedNodeType,
+                    combinePredicates(flippedNodeType, commonPredicates),
```

Comment:
combinePredicates accepts Expression... array. There is no `combinePredicates(type, Expression expression, List<Expression> rest)`. Both arguments need to be `Expression` (not `Expression` and `List<Expression>`)